reframe/core/pipeline.py

```diff
@@ -1655,52 +1656,111 @@ def check_performance(self):
-            for key, ref in self.reference.items():
-                keyparts = key.split(self.reference.scope_separator)
-                system = keyparts[0]
-                varname = keyparts[-1]
-                unit = ref[3]
-                variables.add((varname, unit))
-                if system == '*':
-                    has_default = True
-                    break
-
-            if not has_default:
-                if not variables:
-                    # If empty, it means that self.reference was empty, so try
-                    # to infer their name from perf_patterns
-                    variables = {(name, None)
-                                 for name in self.perf_patterns.keys()}
-
-                for var in variables:
-                    name, unit = var
-                    ref_tuple = (0, None, None, unit)
-                    self.reference.update({'*': {name: ref_tuple}})
-
-            # We first evaluate and log all performance values and then we
-            # check them against the reference. This way we always log them
-            # even if the don't meet the reference.
-            for tag, expr in self.perf_patterns.items():
-                value = sn.evaluate(expr)
-                key = '%s:%s' % (self._current_partition.fullname, tag)
-                if key not in self.reference:
-                    raise SanityError(
-                        "tag `%s' not resolved in references for `%s'" %
-                        (tag, self._current_partition.fullname))
+            if hasattr(self, '_rfm_perf_vars') or self._rfm_perf_fns:
+                if hasattr(self, 'perf_patterns'):
+                    raise ReframeSyntaxError(
+                        f"assigning a value to 'perf_pattenrs' conflicts ",
+                        f"with using the 'performance_variables' decorator ",
+                        f"(class {self.__class__.__qualname__})"
+                    )
 
-                self._perfvalues[key] = (value, *self.reference[key])
-                self._perf_logger.log_performance(logging.INFO, tag, value,
-                                                  *self.reference[key])
+                # Build the performance dict
+                if not hasattr(self, '_rfm_perf_vars'):
+                    self.perf_patterns = {}
+                    for fn in self._rfm_perf_fns:
+                        self.perf_patterns[fn._rfm_perf_name] = fn(self)
 
+                else:
+                    self.perf_patterns = self._rfm_perf_vars()
+
+                # Log the performance variables
+                self._setup_perf_logging()
+                for tag, expr in self.perf_patterns.items():
+                    try:
+                        value, unit = sn.evaluate(expr)
+                        unit = getattr(unit, '_rfm_perf_units', None)
+                        if not unit:
+                            raise TypeError('Trigger the TypeError below')
+
+                    except TypeError:
```

Comment:
This pattern is a bit weird here. We get a  unit that actually may not contain a unit?